nix/tests.nix

```diff
@@ -0,0 +1,57 @@
+{ lib, pkgs }:
+let
+  # TODO: specify project/service name globally
+  application = "web-security-tracker";
+  defaults = {
+    # XXX: we have to duplicate this here despite it being defined in the service module, otherwise the test framework will error out
+    nixpkgs.overlays = lib.mkForce [ (import ./overlay.nix) ];
+    documentation.enable = lib.mkDefault false;
+
+    virtualisation = {
+      memorySize = 2048;
+      cores = 2;
+    };
+
+    services.${application} = {
+      enable = true;
+      production = false;
+      restart = "no"; # fail fast
+      domain = "example.org";
+      secrets = {
+        SECRET_KEY = pkgs.writeText "SECRET_KEY" "secret";
+        GH_CLIENT_ID = pkgs.writeText "gh_client" "bonjour";
+        GH_SECRET = pkgs.writeText "gh_secret" "secret";
+        GH_WEBHOOK_SECRET = pkgs.writeText "gh_secret" "webhook-secret";
+      };
+    };
+  };
+in
+lib.mapAttrs (name: test: pkgs.testers.runNixOSTest (test // { inherit name defaults; })) {
+  application-tests = {
```

Comment:
That's excessively overkill to start 3 VMs to do 3 HTTP requests, no?
You should bundle the test script code and use `subTest`.